WMF Framework/LocationManager.swift

```diff
@@ -0,0 +1,255 @@
+
+    /// The latest known heading.
+    public private(set) lazy var heading: CLHeading? = self.locationManager.heading
+
+    /// Returns `true` when the location and heading monitoring is active.
+    public private(set) var isUpdating = false
+
+    /// Set the delegate if you want to receive location and heading updates.
+    public weak var delegate: LocationManagerDelegate?
+
+    /// Returns the current locationManager permission state.
+    public var authorizationStatus: CLAuthorizationStatus { type(of: locationManager).authorizationStatus() }
+
+    /// Starts monitoring location and heading updates.
+    public func startMonitoringLocation() {
+        
+        guard isUpdating == false else {
+            return
+        }
+        
+        guard authorizationStatus != .notDetermined else {
+            authorize(success: startMonitoringLocation)
+            DDLogVerbose("LocationManager - skip monitoring location because status is \(authorizationStatus.rawValue).")
+            return
+        }
+
+        guard authorizationStatus.isAuthorized else {
+            // Start monitoring location in case the authorization status ever changes to authorized.
+            authorizedCompletion = startMonitoringLocation
+            return
+        }
+
+        locationManager.startUpdatingLocation()
+        startUpdatingHeading()
+        isUpdating = true
+        DDLogDebug("LocationManager - did start updating location & heading.")
+    }
+
+    /// Stops monitoring location and heading updates.
+    public func stopMonitoringLocation() {
+        locationManager.stopUpdatingLocation()
+        stopUpdatingHeading()
+        isUpdating = false
+        DDLogDebug("LocationManager - did stop updating location & heading.")
+    }
+  
+    
+    /// Creates a new instance of `LocationManager`.
+    ///
+    /// - Parameters:
+    ///   - locationManager: If needed, provide a custom `CLLocationManager` instance. The default
+    ///   value creates a new instance.
+    ///   - device: If needed, provide a custom `UIDevice` instance. The default value is `UIDevice.current`.
+    ///   - configuration: The configuration used for configuring `CLLocationManager`. The default value
+    ///   is the `.fine` preset.
+    ///
+    public init(
+        locationManager: CLLocationManager = .init(),
+        device: UIDevice = .current,
+        configuration: LocationManagerConfiguration = .fine
+    ) {
+        assert(Thread.isMainThread, "Location managers must be created on the main thread")
+
+        locationManager.distanceFilter = configuration.filter
+        locationManager.desiredAccuracy = configuration.accuracy
+        locationManager.activityType = configuration.activityType
+        self.locationManager = locationManager
+        self.device = device
+        super.init()
+        locationManager.delegate = self
+    }
+
+    deinit {
+        stopMonitoringLocation()
+        locationManager.delegate = nil
```

Comment:
What do you think about also `nil`'ing this closure out here too? Probably not a big deal, but maybe worth it because we aren't in full control of the authorization lifecycle.

```suggestion
        authorizedCompletion = nil
        locationManager.delegate = nil
```